Add boxed fallback

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26719 Tests for fallback boxed dispatch.
* **#26368 Add boxed fallback**

A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

ezyang committed:
Add boxed fallback

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
This compiles, no pushing on "Add boxed fallback"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
This ACTUALLY compiles, no pushing on "Add boxed fallback"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Closer, now at Tensor& handling on "Add boxed fallback"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

bwasti commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +136,97 @@ class CAFFE2_API ATenDispatch {
+    return (*unboxed_fn)(std::forward<Args>(args)...);
+  }
+
+  auto* unboxed_fallback_fn = reinterpret_cast<FuncType*>(function_table_[static_cast<int64_t>(TensorTypeId::UndefinedTensorId)]);
```
We need the fallback(highest_tid) to be attempted before using `UndefinedTensorId` implementations, which by definition have a lower priority.

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -64,46 +64,49 @@ namespace detail {
     }
   };
 
+  // NB: No universal forwarding
```
comment not needed anymore since you now take `const&`.

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +135,97 @@ class CAFFE2_API ATenDispatch {
+
  private:
   std::unordered_map<std::string, ATenOpTable> op_tables_;
+  FallbackBoxedFunction* boxed_fallback_table_[static_cast<int64_t>(TensorTypeId::NumTensorIds)] = {nullptr};
```
nit: `std::array` is preferrable to C arrays.

ezyang replied:
You gonna delete this code soon right? I did a C array for consistency with the other array in this file.

smessmer commented on aten/src/ATen/test/extension_backend_test.cpp:
```diff
@@ -62,3 +64,12 @@ TEST(BackendExtensionTest, TestRegisterOp) {
+}
+
+TEST(BackendExtensionTest, TestBoxedFallback) {
+  globalATenDispatch().registerFallbackBoxedOp(TensorTypeId::XLATensorId, &generic_override);
```
note: extension backends already switched away from `globalATenDispatch` to `torch::RegisterOperators`

ezyang replied:
Yes this is a good idea.

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +135,97 @@ class CAFFE2_API ATenDispatch {
+    supports_boxed_fallback<Result, Args...>::value,
+    std::nullptr_t
+  > = nullptr>
+Result callBoxedFallback(const char* schema, const FallbackBoxedFunction* boxed_fallback_fn, Args&&... args) {
```
Probably `Args...` instead of `Args&&...`

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +135,97 @@ class CAFFE2_API ATenDispatch {
+    !supports_boxed_fallback<Result, Args...>::value,
+    std::nullptr_t
+  > = nullptr>
+Result callBoxedFallback(const char* schema, const FallbackBoxedFunction* boxed_fallback_fn, Args&&... args) {
```
Probably `Args...` instead of `Args&&...`

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +135,97 @@ class CAFFE2_API ATenDispatch {
+  torch::jit::Stack stack;
+  torch::jit::push(stack, std::forward<Args>(args)...);
+  boxed_fallback_fn(schema, &stack);
+  return torch::jit::pop(stack).to<Result>();
```
maybe assert that there is only one element on the stack? Just to be safe...

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +135,97 @@ class CAFFE2_API ATenDispatch {
+
+  // We need to do a if statement on box fallback support AS WELL
+  // as SFINAE.
+  //    - If statement: because some functions return void, and so
```
I find this explanation confusing. It sounds like the if statement and the SFINAE check different things, but they actually check exactly the same thing. It sounds like the if statement makes sure this works even if it's "return void", but that's actually not true. Since it's not an `if constexpr`, the compiler will still compile both branches, no matter the condition. The reason "return void" compiles fine is because that's an ok thing to do as long as the function itself returns void.

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
namedtensor macrofix on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Include the correct named tensor header on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

ezyang committed:
Update on "Add boxed fallback"


A boxed fallback function is a per-TensorTypeId generic function that is called if there is no operator specific function (either for the TensorTypeId, or the UndefinedTensorId fallback for that operator).

The outline of the patch:
* Add a new `boxed_fallback_table_` in the top-level ATenDispatch struct (this is NOT per operator) for storing boxed fallback functions. Boxed fallback functions provisionally have the interface `void(const char* schema, torch::jit::Stack*)`; we expect to replace the `const char* schema` with something more detailed in the near future.
* Boxing and unboxing is not supported for all arguments, as IValue doesn't cover the full set of types that are in ATen. The `supports_boxed_fallback` type trait tests if boxing is possible. The list of exclusions here was experimentally generated. One notable exclusion is that we cannot handle any mutable functions, as they return a `Tensor&` which we cannot conveniently manufacture from a Tensor on an IValue stack.
* The actual boxed fallback is handled in two phases. First, we do a (compile time) test to see if boxing/unboxing is supported. If it is, we do a runtime test to check if a fallback function is installed. If both conditions are met, we allocate a stack, push our arguments onto it, and then call the boxed fallback function. The return value is expected to be a single argument on the top of the stack; we retrieve it, unpack it, and return it to the user.

At present, there are no tests for this diff.

This diff also makes `multi_dispatch_tensor_type_set` safer by explicitly specifying that it takes all arguments as references. This prevents the function from accidentally moving in rvalue references (this never actually happened because none of the overloads of apply move out their arguments; but they could have.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448555](https://our.internmc.facebook.com/intern/diff/D17448555)

[ghstack-poisoned]

bwasti commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -64,46 +65,59 @@ namespace detail {
+using supports_boxed_fallback =
+  c10::guts::negation<c10::guts::disjunction<
+    std::is_lvalue_reference<Result>,
+    not_ok_to_box<Result>,
```
curious, why the double negatives?

ezyang replied:
no good reason, just how I initially wrote the code. I think one (bad) reason for the negative is because I was trying to avoid saying `std::negation<std::is_same<...,...>>`

bwasti reviewed (approved):
Looks good -- as a test I rebased https://github.com/pytorch/pytorch/pull/25753 onto this

bwasti commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -132,18 +146,104 @@ class CAFFE2_API ATenDispatch {
+  auto* boxed_fallback_fn = globalATenDispatch().getFallbackBoxedOp(tid);
+  if (C10_UNLIKELY(boxed_fallback_fn)) {
+    if (supports_boxed_fallback<Result, Args...>::value) {
+      return callBoxedFallback<Result, Args...>(schema_.c_str(), boxed_fallback_fn, std::forward<Args>(args)...);
```
can we pass in a FunctionSchema rather than the schema string here?

did some light profiling and found dealing with this string on every op invocation is a huge overhead

ezyang replied:
Not surprising :) Yes, FunctionSchema can probably be arranged.

ezyang replied:
Recapping in person conversation: our new plan is to get FunctionSchema out of JIT data structures

bwasti commented on aten/src/ATen/core/stack.h:
```diff
@@ -77,9 +77,21 @@ static inline void pop(Stack& stack, Types&... args) {
 template <typename... Types>
 static inline void push(Stack& stack, Types&&... args) {
-  (void)std::initializer_list<int>{(stack.emplace_back(std::forward<Types>(args)), 0)...};
+  (void)std::initializer_list<int>{(push_one(stack, args), 0)...};
```
do you think we could add a fast path? something like
```
// enable_if all ivalue constructible
stack.reserve(stack.size() + sizeof...(Types));
auto v = { IValue(args)... };
stack.insert(stack.end(), v.begin(), v.end());
```

bwasti replied:
reason: the realloc insert is taking up a non-negligible amount of time in profiling (2%)

ezyang replied:
Yes, I'll do this in a follow up.

ezyang replied:
@bwasti I looked into applying this optimization and I don't think we can safely do it (in fact, the `reserve` in `push_list_elements` is also a bit suspect).  The problem is that reserve can cause quadratic behavior when called in a loop (http://slashslash.info/2013/11/capacity-members-for-vector-reserve/). And in general, the point of putting arguments on a stack is so that we CAN reuse the stack for nested calls without requiring more memory allocations. So reserve is not the right thing here.
